app/account_linking.py

```diff
@@ -0,0 +1,256 @@
+from abc import ABC, abstractmethod
+from arrow import Arrow
+from dataclasses import dataclass
+from enum import Enum
+from typing import Optional
+
+from app.db import Session
+from app.models import PartnerSubscription, Partner, PartnerUser, User
+from app.utils import random_string
+
+
+class SLPlanType(Enum):
+    Free = 1
+    Premium = 2
+
+
+@dataclass
+class SLPlan:
+    type: SLPlanType
+    expiration: Optional[Arrow]
+
+
+@dataclass
+class PartnerLinkRequest:
+    name: str
+    email: str
+    partner_user_id: str
+    plan: SLPlan
+
+
+@dataclass
+class LinkResult:
+    user: User
+    strategy: str
+
+
+class LinkException(Exception):
+    def __init__(self, message: str):
+        self.message = message
+
+
+class AccountAlreadyLinkedToAnotherPartnerException(LinkException):
+    def __init__(self):
+        super().__init__("This account is already linked to another Partner")
+
+
+def set_plan_for_user(user: User, plan: SLPlan, partner: Partner):
+    subs = PartnerSubscription.find_by_user_id(user.id)
+    if plan.type == SLPlanType.Free:
+        # If the plan is Free, check if there is a subscription. If there is, remove it
+        if subs is None:
+            # There was no subscription. We are done
+            return
+        else:
+            # There was a subscription. This is a downgrade. Remove it
+            PartnerSubscription.delete(subs.id)
```

Comment:
For sensitive operation as this one, I think we can add a log which can be handy when investigating later.

Reply:
Added logs to the different branches of this function